### ISSUE TITLE
Fees Discussion

### ISSUE THREAD

**Issue by hassanhabib** [hassanhabib/OtripleS#328]
Hi everyone,
Let's take this week to discuss registration and exam fees functionality in OtripleS.
For a lot of schools - there are fee processing needed to register into school, monthly or by semester fees.
Let's discuss in general what would that look like in a school system. here's some overall guidelines:
1. Not all schools require fees
2. School fees vary based on the entity, there's class fees, exam fees, registration fees and semester fees.
3. We want our schooling system to be pluggable into any payment processing system - it could integrate directly with a bank account, online payment service or some other system

Looking forward to hearing your thoughts on this.

**Comment by viralpandya** [comment]
Other than exam fees, why do we need fees? Most of the public schools are funded by government and they do not need anything other than exam fees and in some cases exam fees is also not needed.

Are we targeting this software to be used by private schools as well?

**Comment by hassanhabib** [comment]
@viralpandya private schools are a nice to have. but mainly public schools. in some public schools kids pay registration fees. they are very small compared to private schools but still.
Also class materials, books and such, they come with fees even in some public schools around the world.

**Comment by eriadhami** [comment]
Do we need to have a 'ExamFee', 'ClassFee', 'RegistrationFee' etc models and then make glue entities with students like 'StudentExamFee' or we can continue with a 'FeeCategory'?

**Comment by hassanhabib** [comment]
@eriadhami yes - exactly. 
We will start with an abstract fee entity then start building those glue entities right after.
I will post the assignments tonight.

**Comment by bilalmehrban** [comment]
I agree with @hassanhabib even public schools does have fee. But it is not as much as compared to the private ones. 

There can be Admission Fees, School bus Fees, Certificates Fee.

An abstract entity is definitely required where we can allow the Admin staff to Add Fees Type. Even the Fees can be different for different level of Standard so we can have some thing like Fee Group as well. Also we can add fees discounts as well i.e. Class Topper Discount, Handicapped Discount, Sibling Discount.

**Comment by hassanhabib** [comment]
Here's an initial idea about the model:

![image](https://user-images.githubusercontent.com/1453985/111581918-5319f480-8777-11eb-9dd9-2251d003f0b3.png)

**Comment by hassanhabib** [comment]
Thank you all for your contributions, here's an overview of our Fee entity model this week:

```csharp
public class Fee {
    public Guid Id {get; set;}
    public string Label {get; set;}
    public decimal Amount {get; set;}

    public Guid CreatedBy {get; set;}
    public User CreatedByUser {get; set; }

    public Guid UpdatedBy {get; set;}
    public User UpdatedByUser {get; set;}

    public DateTimeOffset CreatedDate {get; set;}
    public DateTimeOffset UpdatedDate {get; set;}
}
```
Now, let's talk about Fees:

## Setup 

- Fee Model Setup & Migrations with EF Core @bilalmehrban Tuesday 12:00 AM - 11:59 PM PST

## Broker

- Fee Broker Select By Id function @bilalmehrban by Tuesday 12:00 AM - 11:59 PM PST
- Fee Broker Select All function @bilalmehrban by Tuesday 12:00 AM - 11:59 PM PST
- Fee Broker Insert function @bilalmehrban by Tuesday 12:00 AM - 11:59 PM PST
- Fee Broker Delete function @bilalmehrban by Tuesday 12:00 AM - 11:59 PM PST
- Fee Broker Update function @bilalmehrban by Tuesday 12:00 AM - 11:59 PM PST

## Service

- Fee Service Add function @eriadhami by Wednesday 12:00 AM - 11:59 PM PST
- Fee Service Retrieve by Id function @viralpandya  by Wednesday 12:00 AM - 11:59 PM PST
- Fee Service Retrieve All function @stasnowak by Wednesday 12:00 AM - 11:59 PM PST
- Fee Service Delete by Id function @ZionLloyd05 by Wednesday 12:00 AM - 11:59 PM PST
- Fee Service Modify function @hassanhabib by Wednesday 12:00 AM - 11:59 PM PST

## Controller

- Fee Controller Post endpoint @viralpandya by Thursday 12:00 AM - 11:59 PM PST
- Fee Controller Get by Id endpoint @viralpandya by Thursday 12:00 AM - 11:59 PM PST
- Fee Controller Get All endpoint @viralpandya by Thursday 12:00 AM - 11:59 PM PST
- Fee Controller Delete endpoint @viralpandya by Thursday 12:00 AM - 11:59 PM PST
- Fee Controller Put endpoint @viralpandya by Thursday 12:00 AM - 11:59 PM PST


## Acceptance Tests

- Fee Post Acceptance Test @viralpandya by Friday 12:00 AM - 11:59 PM PST
- Fee Get All Acceptance Test @viralpandya by Friday 12:00 AM - 11:59 PM PST
- Fee Put Acceptance Test @viralpandya by Friday 12:00 AM - 11:59 PM PST
- Fee Delete Acceptance Test @viralpandya by Friday 12:00 AM - 11:59 PM PST

## Contribution Rules:
1. If you have a service task, make sure you commit a failing test, then commit making it pass and so on until your feature is complete (watch this video for reference: https://youtu.be/FLsHIDe3cNs)
2. All CRUD operations now have similar functionality in the AssignmentService, Assignment and Assignment Services, Brokers and Controllers - please follow the same pattern and guidelines.
3. This WIP Coding Standard https://github.com/hassanhabib/CSharpCodingStandard will be our guidance, it is a work in progress but it should give you a pretty good idea bout how our code should look like from guidelines and standardization standpoint.
4. Do not start your task before the designated time and make sure you finalize it at the expected time so others can build on top of your work.

If you haven't had an Fee this week please make sure you engage in discussion for next week so I know you are active in this repo and you are going to be able to contribute to the project.

As soon as all these tasks are done, this issue will be closed.

Thank you all from the bottom of my heart <3 for your contributions, if you are stuck or unsure how to do a task, reach out to me literally on any social media platform you like, LinkedIn, Facebook, Whatsapp, Twitter or even Instagram or just ping me on gitter and I will be sure to response within 2 - 4 hours max.

**Comment by bilalmehrban** [comment]
I am going to do my task today as I will not be available on Tuesday.

**Comment by eriadhami** [comment]
Sorry guys but i can not finish my task before Thursday 05:00 AM PST

**Comment by ZionLloyd05** [comment]
Hello, I'm done with my feature but I noticed that on running the tests immediately I pulled from the master branch, the AssignmentAttachmentAcceptanceApiTests are failing, trying to check it out is the cause of delay.

I would appreciate if anyone checks it out as well and confirm that I'm not mistaken.

![image](https://user-images.githubusercontent.com/40478221/112713821-69425600-8ed7-11eb-810e-22ee39409912.png)

**Comment by viralpandya** [comment]
Checking that just now.

**Comment by viralpandya** [comment]
@ZionLloyd05 : you ca fix acceptance test later. Complete your workitem and create another PR for fixing acceptance tests.

I am not having that issue. So, once all the PRs merged in, you can check again if issue persists.

**Comment by ZionLloyd05** [comment]
Ok, are you suggesting that I should hold on for a merge and pull an updated version or submit the PR for my task, considering that the test issue still persists?

**Comment by ZionLloyd05** [comment]
Ok, are you suggesting that I should hold on for a merge and pull an updated version or submit the PR for my task, considering that the test issue still persists?
